### PR TITLE
Add vector agent config to stargate

### LINKED ISSUE #789: Deploy Vector agent for Reaper metrics scraping
Following up on #775, we'd need to deploy Vector as a sidecar in the Reaper container in order to scrape its metrics and forward them based on a user provided configuration.

The api changes will be done in #775 in the telemetry spec.

### LINKED ISSUE #790: Deploy Vector agent to scrape Stargate metrics
Following up on https://github.com/k8ssandra/k8ssandra-operator/issues/775, we'd need to deploy Vector as a sidecar in the Stargate container in order to scrape its metrics and forward them based on a user provided configuration.

The api changes will be done in https://github.com/k8ssandra/k8ssandra-operator/issues/775 in the telemetry spec.

### PR DESCRIPTION
**Pull request by emerkle826**
**What this PR does**:
Adds ability to configure Vector as a sidecar for Stargate and Reaper

**Which issue(s) this PR fixes**:
Fixes #790 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: controllers/stargate/vector.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	vectorToml := new(bytes.Buffer)
+	err = t.Execute(vectorToml, config)
+	if err != nil {
+		panic(err)
+	}
+
+	return vectorToml.String(), nil
+}
+
+func InjectVectorAgentForStargate(stargate *api.Stargate, deployments map[string]appsv1.Deployment, dcName, clustername string, logger logr.Logger) error {
```
[the remark this answers is not in the document]
Comment (adutra) [reply controllers/stargate/vector.go:164]: This function always returns a nil error, so you could remove the `error` return parameter altogether.

File: controllers/stargate/stargate_controller.go [hunk 2]
```diff
@@ -265,6 +265,19 @@ func (r *StargateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
 
+	// create Vector configmap
+	if vectorReconcileResult, err := r.reconcileVector(ctx, *stargate, actualDc, r.Client, logger); err != nil {
+		return vectorReconcileResult, err
+	} else if vectorReconcileResult.Requeue {
+		return vectorReconcileResult, nil
+	}
+
+	// inject Vector
+	if err := InjectVectorAgentForStargate(stargate, desiredDeployments, actualDc.Name, actualDc.Spec.ClusterName, logger); err != nil {
```
Comment (adutra) [inline controllers/stargate/stargate_controller.go:276]: As stated in another comment, this function doesn't return any error, so we could remove the error handling code here.

**Review by adutra — CHANGES_REQUESTED** [review]

File: pkg/stargate/config.go [hunk 3]
```diff
@@ -58,6 +59,26 @@ func CreateStargateConfigMap(namespace, cassandraYaml, stargateCqlYaml string, d
...
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      stargateapi.VectorAgentConfigMapNameStargate(dc.Spec.ClusterName, dc.Name),
+			Namespace: namespace,
+			Labels: map[string]string{
+				k8ssandra.NameLabel:                      k8ssandra.NameLabelValue,
+				k8ssandra.PartOfLabel:                    k8ssandra.PartOfLabelValue,
+				k8ssandra.ComponentLabel:                 k8ssandra.ComponentLabelValueStargate,
+				k8ssandra.CreatedByLabel:                 k8ssandra.CreatedByLabelValueK8ssandraClusterController,
+				k8ssandra.K8ssandraClusterNameLabel:      dc.Labels[k8ssandra.K8ssandraClusterNameLabel],
```
[the remark this answers is not in the document]
Comment (emerkle826) [reply pkg/stargate/config.go:72]: @adutra I have update the code so that the Stargate Vector ConfigMap's controller reference is set to the Stargate object here:
https://github.com/k8ssandra/k8ssandra-operator/pull/819/files#diff-2e5c443ca862a5be41c3c80f344d80d6e0b83c0084c8724a0364330699d5df6fR55

which is the same thing we do for the Stargate ConfigMap here:
https://github.com/k8ssandra/k8ssandra-operator/blob/5040b3fb075c09312f57bfbfd49c9b0c86f3b63b/controllers/stargate/stargate_controller.go#L437

File: apis/stargate/v1alpha1/stargate_types.go [hunk 4]
```diff
@@ -416,3 +418,7 @@ type StargateList struct {
 func init() {
 	SchemeBuilder.Register(&Stargate{}, &StargateList{})
 }
+
+func VectorAgentConfigMapNameStargate(clusterName, dcName string) string {
```
[the remark this answers is not in the document]
Comment (emerkle826) [reply apis/stargate/v1alpha1/stargate_types.go]: I have moved the function to `pkg/stargate` as I use it in `pkg/stargate` as well as `controller/stargate`

File: controllers/stargate/vector.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+  [sinks.console.encoding]
+  codec = "json"`
+
+	if telemetrySpec.Vector.Components != nil {
+		// Vector components are provided in the Telemetry spec, build the Vector sink config from them
+		vectorConfigToml = telemetry.BuildCustomVectorToml(telemetrySpec)
+	}
+
+	var scrapeInterval int32 = telemetry.DefaultScrapeInterval
```
[the remark this answers is not in the document]
Comment (emerkle826) [reply controllers/stargate/vector.go:130]: I've created a ticket for this: https://github.com/k8ssandra/k8ssandra-operator/issues/826

File: controllers/stargate/stargate_controller.go [hunk 6]
```diff
@@ -265,6 +265,16 @@ func (r *StargateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
 
+	// create Vector configmap
+	if vectorReconcileResult, err := r.reconcileVector(ctx, *stargate, actualDc, r.Client, logger); err != nil {
+		return vectorReconcileResult, err
+	} else if vectorReconcileResult.Requeue {
+		return vectorReconcileResult, nil
+	}
+
+	// inject Vector
+	injectVectorAgentForStargate(stargate, desiredDeployments, actualDc.Name, actualDc.Spec.ClusterName, logger)
```
Comment (adutra) [inline controllers/stargate/stargate_controller.go:276]: You need to call this function from inside `stargateutil.NewDeployments` because the hash annotation is being computed there.

File: controllers/stargate/vector.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+		if err := deleteConfigMapIfExists(ctx, remoteClient, configMapKey, dcLogger); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	dcLogger.Info("Vector Agent ConfigMap successfully reconciled")
+	return ctrl.Result{}, nil
+}
+
+func deleteConfigMapIfExists(ctx context.Context, remoteClient client.Client, configMapKey client.ObjectKey, logger logr.Logger) error {
```
Comment (adutra) [inline controllers/stargate/vector.go:99]: We have a similar function already for K8ssandraCluster config maps, maybe we could extract this into a generic `deleteObjectIfExists`?

File: pkg/nodeconfig/mount.go [hunk 8]
```diff
@@ -3,7 +3,7 @@ package nodeconfig
 import (
 	"github.com/k8ssandra/cass-operator/pkg/reconciliation"
 	"github.com/k8ssandra/k8ssandra-operator/pkg/cassandra"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (adutra) [inline pkg/nodeconfig/mount.go:6]: Unnecessary aliasing? Besides we usually use an alias composed of the last part before the version + the version itself, so: `corev1`.

**Review by adutra — CHANGES_REQUESTED** [review]

File: controllers/stargate/stargate_controller.go (hunk 6 above)
Comment (emerkle826) [reply controllers/stargate/stargate_controller.go:276]: Yes, you are correct. That is what was causing my changes to not get picked up.

File: pkg/nodeconfig/mount.go (hunk 8 above)
Comment (emerkle826) [reply pkg/nodeconfig/mount.go:6]: Good catch. I blame my IDE :)

File: controllers/stargate/vector.go (hunk 7 above)
Comment (emerkle826) [reply controllers/stargate/vector.go:99]: Are you referring to the per_node config deletion here:
https://github.com/k8ssandra/k8ssandra-operator/blob/main/controllers/k8ssandra/per_node_config.go#L102-L111

File: controllers/stargate/stargate_controller.go [hunk 9]
```diff
@@ -195,7 +195,14 @@ func (r *StargateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 	}
 
 	// Compute the desired deployments
-	desiredDeployments := stargateutil.NewDeployments(stargate, actualDc)
+	desiredDeployments := stargateutil.NewDeployments(stargate, actualDc, logger)
```
Comment (adutra) [inline controllers/stargate/stargate_controller.go:198]: Not a big deal, but maybe you could reconcile the Vector config map first, then call this function. This way if there is an error with the former, we don't lose time calling the latter.

File: controllers/stargate/vector.go (hunk 7 above)
Comment (adutra) [reply controllers/stargate/vector.go:93]: Hmm nevermind I think the function was deleted or refactored recently.

File: controllers/stargate/vector.go [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+			}
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+		}
+	} else {
+		if err := deleteConfigMapIfExists(ctx, remoteClient, configMapKey, dcLogger); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	dcLogger.Info("Stargate Vector Agent ConfigMap successfully reconciled")
```
Comment (adutra) [inline controllers/stargate/vector.go:89]: Nit: this will print this message even if Vector is disabled and there is no config map to reconcile.

File: pkg/stargate/config.go [hunk 11]
```diff
@@ -10,7 +10,9 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-const CqlConfigName = "stargate-cql.yaml"
+const (
```
Comment (adutra) [inline pkg/stargate/config.go:13]: Change is not really necessary.

File: pkg/stargate/vector_test.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/k8ssandra/k8ssandra-operator/pkg/vector"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/utils/pointer"
+)
+
+// InjectCassandraVectorAgent adds the Vector agent container to the Cassandra pods.
+// If the Vector agent is already present, it is not added again.
+func TestConfigureStargate(t *testing.T) {
```
Comment (adutra) [inline pkg/stargate/vector_test.go:19]: Nit: the test name should rather be something like `TestConfigureVector`. Also, the comment above seems unrelated.

**Review by adutra — APPROVED** [review]

File: controllers/stargate/vector.go (hunk 10 above)
Comment (emerkle826) [reply controllers/stargate/vector.go:89]: Yes, that was a debug message I was using earlier. I'll get rid of it.
